Add `HasSlug` mixin

jcredding opened the pull request:
This adds a `HasSlug` mixin that provides automatic slugging to
active record models. This allows creating readable slugs for
easily.

The mixin adds a `has_slug` class method when included. This can
be used to add the automatic slugging behavior and configure it as
needed. The `has_slug` requires a source proc for configuring
how the slug is generated. It uses this to generate the initial
slug.

Once `has_slug` has been called, it adds validations and callbacks.
It adds a validates presence on update for the configured slug
attribute. Its on update because slugs are generated after the
model has been saved (so the `id` can be used). It also adds a
validates uniqueness of for the slug attribute. This can be
further customized by specifying a unique scope option when
calling `has_slug`. It adds an after create and after update
callback. Both of these methods generate a slug. This is done on
create and update to initially generate the slug and also to make
sure the slug stays a valid slug. If the slug attribute is already
set, the callbacks will attempt to slug the existing string to
make sure it stays a valid slug. Otherwise, the source proc is
used to get an initial string.

This also updates the `UseDbDefault` mixin to prefix its config
method with `ardb`. This should've been done when it was
implemented but was accidentally missed.

@kellyredding - Ready for review.

kellyredding commented on lib/ardb/has_slug.rb:
```diff
@@ -0,0 +1,101 @@
+
+      def has_slug_allow_underscores
+        @ardb_has_slug_allow_underscores
+      end
```
@jcredding really don't love adding all these class methods.  Other implementation didn't do this.

kellyredding commented:
@jcredding we can discuss more in person, but I prefer the `Configuration` implementation we used previously in our apps.  It results in less "method pollution".

Another (admittedly more subjective reason) is that it seems to encapsulate the logic better.  I felt (key word: me "felt") this implementation splays out its logic and handling all over the record.  In this implementation you have to add many ivars/methods to store the configuration state and in the previous implementation the configuration was the only thing created and it wasn't even stored on the singleton (was implicitly stored in the callbacks as a reference).  Plus you are forced to use really long variable names when the configuration implicitly namespaced things for you.

Maybe you can outline some of the reasons why you went away from the configuration approach?  We can discuss in person later of course.

kellyredding commented:
@jcredding sorry to come off so negative on this.  I like how this implementation lets you customize the attribute name.

kellyredding commented on lib/ardb/has_slug.rb:
```diff
@@ -0,0 +1,101 @@
+      private
+
+      def reset_slug
+        self.send("#{self.class.has_slug_attribute}=", nil)
```
@jcredding ahh, maybe this is why you went away from config implementation?  You need access to this config?

kellyredding commented:
@jcredding ok I am seeing (maybe?) a bit more why you went away from config implementation.  I'd like to hear more from you though.  If my assumptions are correct, there may be a way we can achieve this feature-set and not pollute the record so much.  Anyway, we can discuss.

(notes to self)

Option #1: store config values as a single obj on class (struct/hash something).  only requires one `ardb_has_slug_config` (or whatever) class method/ivar.

Option #2 (preferred by me): use previous config implementation to bind config to record (`add_validations`, `add_callbacks`) like before.  Need an additional method `add_reset_slug_method`.  Could combine all this logic into a single `bind` method or something.

kellyredding commented:
@jcredding also thanks for tackling this in the first place - don't take my negative comments as a lack of appreciation :).

jcredding commented:
@kellyredding - As we discussed, I went with this implementation for a few reasons:

* This is faster and more importantly uses less memory than the configuration and closures implementation did
* I realize this is subjective but I felt this is a simpler and more straightforward approach. Relying on ruby closures is cool but is not the easiest to follow pattern.
* I also wanted to avoid the need for a spy, so at the minimum I would want to expose the configuration as a class method.

The speed and memory is my biggest reason but all of these are relatively minor and I think there are trade-offs for all the implementations.

I updated this to first namespace the methods with `ardb_` (I also updated `UseDbDefault` to do this) and I switched to option 1. There is now just one class method that returns a hash that has all the values. This performs as good as (or better) than my original implementation but doesn't pollute the class with lots of methods. As you mentioned, it also scales nicely if there are more options later on.

kellyredding commented on lib/ardb/has_slug.rb:
```diff
@@ -0,0 +1,93 @@
+          :allow_underscores => !!options[:allow_underscores]
+        })
+
+        # can't always b/c of slug is written `after_save`
```
@jcredding this comment seems off - "b/c *of* slug is written"?

kellyredding commented:
@jcredding thanks for that feedback - that all makes sense.  Now that the 'non-user-facing' class methods have been namespaced, I'm OK with taking this approach - especially in light of the performance improvements.

:-1:
